book/optimization.qmd

```diff
@@ -772,12 +592,404 @@ It is therefore important to ensure that the performance of a tuned model is *al
+search_space = ps(
+  cost = p_dbl(-1, 1, trafo = function(x) 10^x),
+  kernel = p_fct(c("polynomial", "radial")),
+  .extra_trafo = function(x, param_set) {
+    if (x$kernel == "polynomial") {
+      x$cost = x$cost + 2
+    }
+    x
+  }
+)
+```
+
+```{r optimization-040}
+#| echo: false
+#| label: fig-technical-paradox-trafo
+#| fig-cap: Design grid for tuning a SVM. The resolution is 5, the cost parameter is logaithmically transformed when points with a `kernel` equal to `"polynomial"` are shifted to the right by a value of 2.
+dat3 = rbindlist(generate_design_grid(search_space, 5)$transpose())
+theme_set(theme_minimal())
+ggplot(data = dat3, aes(cost, kernel)) +
+    geom_point()
+```
+
+The available types of search space parameters are limited: continuous, integer, discrete, and logical scalars.
+There are many machine learning algorithms, however, that take parameters of other types, for example vectors or functions.
+These can not be defined in a search space `r ref("ParamSet")`, and they are often given as `r ref("ParamUty")` in the `r ref("Learner")`'s `r ref("ParamSet")`.^[Recall that a parameter of type `r ref("ParamUty")` cannot be used in a search space because it is never bounded.]
+When trying to tune over these hyperparameters, it is necessary to perform a Transformation that changes the type of the parameter.
+
+An example is the `class.weights` parameter of the [Support Vector Machine](https://machinelearningmastery.com/cost-sensitive-svm-for-imbalanced-classification/) (SVM), which takes a named vector of class weights with one entry for each target class.
+The transformation that would tune `class.weights` for the `r ref("mlr_tasks_spam", text = "spam")` dataset could be:
+
+```{r optimization-041}
+search_space = ps(
+  class.weights = p_dbl(lower = 0.1, upper = 0.9,
+    trafo = function(x) c(spam = x, nonspam = 1 - x))
+)
+```
+
+A common use-case is the necessity to specify a list of values that should all be tried (or sampled from).
+It may be the case that a hyperparameter accepts function objects as values and a certain list of functions should be tried.
+Or it may be that a choice of special numeric values should be tried.
+For this, the `r ref("p_fct")` constructor's `level` argument may be a value that is not a `character` vector, but something else.
+If, for example, only the values 0.1, 3, and 10 should be tried for the `cost` parameter, even when doing random search, then the following search space would achieve that:
+
+```{r optimization-042}
+search_space = ps(
+  cost = p_fct(c(0.1, 3, 10)),
+  kernel = p_fct(c("polynomial", "radial"))
+)
+```
+
+This is equivalent to the following:
+```{r optimization-043}
+search_space = ps(
+  cost   = p_fct(c("0.1", "3", "10"),
+    trafo = function(x) list(`0.1` = 0.1, `3` = 3, `10` = 10)[[x]]),
+  kernel = p_fct(c("polynomial", "radial"))
+)
+```
+
+This may seem silly, but makes sense when considering that factorial tuning parameters are always `character` values:
```

Comment:
Never say "this may seem silly" about your own code

Reply:
@sebffischer